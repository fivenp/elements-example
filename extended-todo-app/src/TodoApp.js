import React, { Component } from 'react'
import { ColorPalette } from '@allthings/colors'
import Icon from '@allthings/elements/atoms/Icon'
import Inset from '@allthings/elements/atoms/Inset'
import Text from '@allthings/elements/atoms/Text'
import View from '@allthings/elements/atoms/View'
import Button from '@allthings/elements/molecules/Button'
import Card from '@allthings/elements/molecules/Card'
import TextInput from '@allthings/elements/molecules/TextInput'
import GroupedCardList from '@allthings/elements/organisms/CardList/GroupedCardList'
import TitleBar from '@allthings/elements/organisms/TitleBar'
import ResourceProvider from '@allthings/elements/behaviour/ResourceProvider'
import ThemeProvider from '@allthings/elements/behaviour/ThemeProvider'
import TodoItem from './TodoItem'
import { css } from 'glamor'

const styles = {
  create: css({
    ':hover': { cursor: 'pointer' },
  }),
  inset: css({
    width: '93%',
  }),
  new: css({
    padding: 0,
    width: '100%',
    display: 'flex',
  }),
  iconStyle: css({
    display: 'inline-block',
    marginRight: '10px',
    position: 'absolute',
  }),
  titleBar: css({
    padding: '10px 15px 10px 0px',
  }),
  titleText: css({
    display: 'inline-block',
    padding: '0 35px',
    position: 'relative',
  }),
  buttonDiv: css({
    textAlign: 'center',
  }),
  button: css({
    margin: '3px',
  }),
  textInput: css({
    display: 'inline-block',
    width: '100%',
    ':focus': { outline: 'none' },
  }),
  add: css({
    display: 'inline-block',
    float: 'right',
    margin: '15px 10px 15px 0px',
  }),
}

const buttonOpacity = 'rgba(232, 76, 61, 0.5)'

const DemoTheme = {
  primary: ColorPalette.red,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
}

class App extends Component {
  state = {
    todos: [
      {
        id: 0,
        text: 'Search for cocktail recipes for the party',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
      {
        id: 1,
        text: 'Create nice invitation cards',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
      {
        id: 2,
        text: 'Ask some people to bring some finger food',
        done: false,
        doubleClicked: false,
        checkmark: 'flex',
        iconOpen: false,
      },
    ],
    filter: 'all',
    textDisable: true,
    newText: '',
  }

  updateInput = event => {
    this.setState({ newText: this.textInput.input.value })
  }

  componentDidMount = () => {
    document.addEventListener('touchstart', this.finishEdit)
    document.addEventListener('mousedown', this.finishEdit)
  }

  handleRemove = id => {
    const newTodo = this.state.todos.slice()
    newTodo.splice(id, 1)
    this.setState({ todos: newTodo })
  }

  handleDoubleClick = id => {
    const newTodo = this.state.todos.slice()
    newTodo[id].doubleClicked = true
    this.setState({ todos: newTodo })
  }

  handleSliding = (id, type) => {
    const newTodo = [...this.state.todos]
    newTodo[id].iconOpen = type === 'open' ? true : false
    this.setState({ todos: newTodo })
  }

  checkmarkClicked = id => {
    const newTodo = [...this.state.todos]
    newTodo[id].done = !this.state.todos[id].done
    this.setState({ todos: newTodo })
  }

  handleKeyPress = event => {
    const text = event.target.value
    if (event.key === 'Enter' && text) {
      this.addTodo()
    } else {
      const textExists = this.textInput.input.value
      if (textExists && this.state.textDisable) {
        this.setState({ textDisable: false })
      } else if (!textExists && !this.state.textDisable) {
        this.setState({ textDisable: true })
      }
    }
  }

  addTodo = () => {
    const text = this.textInput.input.value
    this.setState({
      todos: [
        ...this.state.todos,
        {
          text,
          done: false,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
        },
      ],
      textDisable: true,
    })
    this.setState({ newText: '' })
    this.textInput.input.blur()
  }

  changeFilter = ({ target: { className } }) => {
    this.setState({ filter: className })
  }

  focusInput = () => {
    this.textInput.input.focus()
  }

  doneEditting = (event, id) => {
    const text = event.target.value
    if (event.key === 'Enter' && text) {
      const newTodo = [...this.state.todos]
      newTodo[id] = {
        ...newTodo[id],
        doubleClicked: false,
        text,
        checkmark: 'flex',
        iconOpen: false,
      }
      this.setState({ todos: newTodo })
    }
  }

  finishEdit = event => {
    const oldTodo = [...this.state.todos]
    const eventSrc = event.srcElement
    const { id: eventId, nodeName: eventNode, className: eventClass } = eventSrc
    const newTodo = oldTodo.map((todo, index) => {
      const { doubleClicked, iconOpen } = todo
      if (eventClass === 'icon-div' || eventNode === ('svg' || 'path')) {
        return todo
      } else if ((doubleClicked || iconOpen) && index.toString() === eventId) {
        return todo
      } else if (doubleClicked && eventClass !== 'icon-div') {
        todo = {
          ...todo,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
          text: document.getElementById(index).value,
        }
        return todo
      } else if (iconOpen && eventClass !== 'icon-div') {
        todo = {
          ...todo,
          doubleClicked: false,
          checkmark: 'flex',
          iconOpen: false,
        }
        return todo
      } else {
        return todo
      }
    })
    this.setState({ todos: newTodo })
  }

  render() {
    const { todos, filter, textDisable } = this.state
    const incomplete = todos.reduce(
      (count, todo) => (todo.done ? count : count + 1),
      0
    )
    const incompleteNum =
      incomplete > 0 ? `Incomplete (${incomplete})` : 'Incomplete'

    return (
      <div className="App">
        <ResourceProvider>
          <ThemeProvider theme={DemoTheme}>
            <View>
              <TitleBar {...styles.titleBar}>
                <Inset {...styles.inset}>
                  <Icon
                    name="book-open"
                    size="m"
                    color="white"
                    {...styles.iconStyle}
                  />
                  <Text color="contrast" {...styles.titleText}>
                    My Todos
                  </Text>
                </Inset>
                <Icon
                  name="plus-light"
                  size="m"
                  color="#FFFFFF"
                  onClick={this.focusInput}
                  {...styles.create}
                />
              </TitleBar>
              <Card>
                <div {...styles.new}>
                  <div {...styles.textInput}>
                    <TextInput
                      id="new"
                      name="new"
                      ref={node => (this.textInput = node)}
                      placeholder="Add new"
                      value={this.state.newText}
                      onKeyUp={this.handleKeyPress}
                      onChange={this.updateInput.bind(this)}
                      {...styles.textInput}
                    />
                  </div>
                  <Button
                    id="add-task"
                    onClick={this.addTodo}
                    disabled={textDisable}
                    {...styles.add}
                  >
                    add
                  </Button>
                </div>
                <div {...styles.buttonDiv}>
                  <Button
                    className="all"
                    onClick={this.changeFilter}
                    backgroundColor={filter !== 'all' ? buttonOpacity : ''}
                    {...styles.button}
                  >
                    <span className="all">All</span>
                  </Button>
                  <Button
                    className="incomplete"
                    onClick={this.changeFilter}
                    backgroundColor={
                      filter !== 'incomplete' ? buttonOpacity : ''
                    }
                    {...styles.button}
                  >
                    <span className="incomplete">{incompleteNum}</span>
                  </Button>
                  <Button
                    id="completed-button"
                    className="completed"
                    onClick={this.changeFilter}
                    backgroundColor={
                      filter !== 'completed' ? buttonOpacity : ''
                    }
                    {...styles.button}
                  >
                    <span className="completed">Completed</span>
                  </Button>
                </div>
                <GroupedCardList>
                  {todos.map((todo, index) => {
                    const {
                      done,
                      checkmark,
                      iconOpen,
                      doubleClicked,
                      text,
                    } = todo
                    const thisTodoItem = (
                      <TodoItem
                        handleRemove={this.handleRemove}
                        id={index}
                        done={done}
                        checkmarkDisplay={checkmark}
                        iconOpen={iconOpen}
                        doubleClicked={doubleClicked}
                        key={'checkbox' + index}
                        onDoubleClick={this.handleDoubleClick}
                        onSlideLeft={this.handleSliding}
                        doneEditting={this.doneEditting}
                        checkmarkClicked={this.checkmarkClicked}
                      >
                        {text}
                      </TodoItem>
                    )
                    if (
                      (filter === 'incomplete' && done === false) ||
                      ((filter === 'completed' && done === true) ||
                        filter === 'all')
                    ) {
                      return thisTodoItem
                    }
                  })}
                </GroupedCardList>
              </Card>
            </View>
          </ThemeProvider>
        </ResourceProvider>
      </div>
    )
  }
}

export default App
