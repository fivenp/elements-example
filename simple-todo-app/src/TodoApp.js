import React, { Component } from 'react'
import { ColorPalette } from '@allthings/colors'
import { Card, List, Inset, ResourceProvider, TitleBar, Text, ThemeProvider } from '@allthings/elements'
import TodoItem from './TodoItem'

const DemoTheme = {
  primary: ColorPalette.red,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright
}

class App extends Component {
  state = {
    todos: [
      { id: 1, text: 'Search for cocktail recipes for the party', done: false },
      { id: 2, text: 'Create nice invitation cards', done: false },
      { id: 3, text: 'Ask some people to bring some finger food', done: false }
    ]
  }

  handleClick = id => {
    this.setState(({ todos }) => ({
      todos: todos.map(todo => {
        if (todo.id === id) {
          return {
            ...todo,
            done: !todo.done
          }
        } else {
          return todo
        }
      })
    }))
  }

  render() {
    return (
      <div className="App">
        <ThemeProvider theme={DemoTheme}>
          <ResourceProvider>
            <TitleBar>
              <Inset>
                <Text strong color="white">
                  Todos
                </Text>
              </Inset>
            </TitleBar>
            <Card>
              <List>
                {this.state.todos.map(todo => (
                  <TodoItem
                    onClick={this.handleClick}
                    id={todo.id}
                    done={todo.done}
                  >
                    {todo.text}
                  </TodoItem>
                ))}
              </List>
            </Card>
          </ResourceProvider>
        </ThemeProvider>
      </div>
    )
  }
}

export default App
