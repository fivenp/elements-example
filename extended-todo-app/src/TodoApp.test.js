import React from 'react'
import { shallow, mount, render } from 'enzyme'
import TodoApp from './TodoApp'

describe('TodoApp', () => {
  const wrapper = shallow(<TodoApp />)
  const completeWrapper = wrapper.setState({
    todos: [
      { text: 'Text 1', done: false },
      { text: 'Text 2', done: true },
      { text: 'Text 3', done: true },
    ],
    filter: 'completed',
    newText: 'testing',
  })
  it('should match snapshot (no buttons clicked)', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have one .App div', () => {
    expect(wrapper.find('.App')).toHaveLength(1)
  })

  it('should have one #add-task', () => {
    expect(wrapper.find('#add-task')).toHaveLength(1)
  })

  it('should render correct number of complete', () => {
    expect(completeWrapper.find('TodoItem')).toHaveLength(2)
  })
})

describe('TodoApp buttons', () => {
  const component = mount(<TodoApp />).setState({
    todos: [
      { text: 'Text 1', done: true },
      { text: 'Text 2', done: false },
      { text: 'Text 3', done: false },
    ],
    filter: 'all',
    textDisable: false,
    newText: 'Text 4',
  })

  it('add button should add one TodoItem with correct text and #new TextInput value is cleared', () => {
    component
      .find('#add-task')
      .first()
      .simulate('click')
    expect(component.find('TodoItem')).toHaveLength(4)
    expect(
      component
        .find('TodoItem')
        .at(3)
        .prop('children')
    ).toBe('Text 4')
    expect(component.find('TextInput').prop('value')).toBe('')
  })

  it('filter buttons should filter correctly', () => {
    component
      .find('#completed-button')
      .first()
      .simulate('click')
    expect(component.find('TodoItem')).toHaveLength(1)
  })
})
