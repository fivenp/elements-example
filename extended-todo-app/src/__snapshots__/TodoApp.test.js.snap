// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp should match snapshot (no buttons clicked) 1`] = `
<div
  className="App"
>
  <ResourceProvider
    resourcePath="https://static.allthings.me"
  >
    <ThemeProvider
      theme={
        Object {
          "background": "#f3f5f7",
          "contrast": "#ffffff",
          "disabled": "#95a5a5",
          "primary": "#e84c3d",
          "secondaryText": "#333333",
          "text": "#333333",
          "titleColor": "#333333",
        }
      }
    >
      <View
        alignH="start"
        alignV="stretch"
        fill={false}
        flex="none"
        onRef={[Function]}
      >
        <TitleBar
          data-css-bh3vuo=""
        >
          <Inset
            data-css-14z3si=""
            horizontal={true}
          >
            <Icon
              color="white"
              data-css-xih1gx=""
              name="book-open"
              size="m"
            />
            <Text
              block={true}
              color="contrast"
              data-css-1y7q7mp=""
              size="l"
            >
              My Todos
            </Text>
          </Inset>
          <Icon
            color="#FFFFFF"
            data-css-1eqigeq=""
            name="plus-light"
            onClick={[Function]}
            size="m"
          />
        </TitleBar>
        <Card>
          <div
            data-css-l09ulh=""
          >
            <div
              data-css-r9ebme=""
            >
              <TextInput
                data-css-r9ebme=""
                id="new"
                lines={1}
                name="new"
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="Add new"
                required={false}
                type="text"
                value="testing"
              />
            </div>
            <_class
              data-css-1bn1z22=""
              disabled={true}
              id="add-task"
              onClick={[Function]}
            >
              add
            </_class>
          </div>
          <div
            data-css-1yv2jt6=""
          >
            <_class
              backgroundColor="rgba(232, 76, 61, 0.5)"
              className="all"
              data-css-pg4ma5=""
              onClick={[Function]}
            >
              <span
                className="all"
              >
                All
              </span>
            </_class>
            <_class
              backgroundColor="rgba(232, 76, 61, 0.5)"
              className="incomplete"
              data-css-pg4ma5=""
              onClick={[Function]}
            >
              <span
                className="incomplete"
              >
                Incomplete (1)
              </span>
            </_class>
            <_class
              backgroundColor=""
              className="completed"
              data-css-pg4ma5=""
              id="completed-button"
              onClick={[Function]}
            >
              <span
                className="completed"
              >
                Completed
              </span>
            </_class>
          </div>
          <GroupedCardList>
            <TodoItem
              checkmarkClicked={[Function]}
              done={true}
              doneEditting={[Function]}
              handleRemove={[Function]}
              id={1}
              key="checkbox1"
              onDoubleClick={[Function]}
              onSlideLeft={[Function]}
            >
              Text 2
            </TodoItem>
            <TodoItem
              checkmarkClicked={[Function]}
              done={true}
              doneEditting={[Function]}
              handleRemove={[Function]}
              id={2}
              key="checkbox2"
              onDoubleClick={[Function]}
              onSlideLeft={[Function]}
            >
              Text 3
            </TodoItem>
          </GroupedCardList>
        </Card>
      </View>
    </ThemeProvider>
  </ResourceProvider>
</div>
`;
